### ISSUE TITLE
Discrepancy in width/height of polygons and meshes

### ISSUE THREAD

**Issue by haesleinhuepf** [imagej/imagej-ops#439]
Hey guys,

I just figured out, that there is a discrepancy between how polygons and meshes are generated / interpreted. You can see the issue by comparing X/Y coordinates of vertices of polygons and meshes being retrieved from similar binary masks.

Just an example: imagine a binary mask with 4x4 pixels being positive. The x coordinate of the first positive pixel may be 2 and the last one is at position 5. We do the same with a 4x4x4 cube in 3D. Now we generate a polygon from the 2D mask and a mesh from the 3D mask using ops contour()  and marchingCubes(), respectively. Then we look at the vertices:

- The minimum x-coordinate of the polygon is 2, the maximum is 5. Its width appears to be 5 - 2 = 3
- The minimum x-coordinate of the mesh is 1.5, the maximum is 5.5. Its width appears to be 5.5 - 1.5 = 4.

Both results appear to be right from some point of view and both appear wrong from another point of view. And furthermore, I could imagine, that the polygon ranges from 2 to 6, because it would then be indeed outlining the binary mask. The related question is as old as image processing itself: Is the position of a pixel on the pixels top left corner or in its center? Independent from the answer to this question: I hope we all agree that this issue should to be solved at least in a way that polygons and meshes are treated equally.

For those who are interested in details, there is an implementation for showing the issue available online:
https://github.com/haesleinhuepf/OpsPolygonTest/blob/master/src/test/java/de/mpicbg/scf/labelhandling/OpsBoundingIntervalsTest.java

Cheers,
Robert

**Comment by ctrueden** [comment]
> Is the position of a pixel on the pixels top left corner or in its center?

Yep, this is the age-old "what are the coordinates of a sample" discussion. [Here](https://groups.google.com/d/msg/fiji-devel/AdeqZKffIUU/9SoisoaivWwJ) is one old thread discussing some of the issues, back when ImageJ2 and ImgLib were first getting started.

There is no such thing as the "top left corner" of a pixel. [Pixels are not little squares](http://imagej.net/Principles#Pixels_are_not_little_squares).

In ImgLib2 terms, the "top left" sample of your binary mask is at (0, 0). The "top right" is at (0, w-1). The "bottom right" is at (w-1, h-1). Same applies to 3D and beyond.

> I hope we all agree that this issue should to be solved at least in a way that polygons and meshes are treated equally.

Definitely.

> The minimum x-coordinate of the polygon is 2, the maximum is 5. Its width appears to be 5 - 2 = 3

To me, this is correct. The binary mask has four positive samples along width, and four positive samples along height. We treat each sample as a point here, ignoring issues of PSF et. al. Hence, the bounding polygon "hugs" the samples tightly, meaning it has width 3 and height 3.

> I could imagine, that the polygon ranges from 2 to 6, because it would then be indeed outlining the binary mask.

This would mean the polygon hugs the top-left side tightly, but leaves a padding of width 1 on the bottom-right. That seems inconsistent to me. However, I think Java AWT actually works that way. That said, consistently with AWT is not one of our design requirements here.

**Comment by haesleinhuepf** [comment]
> Hence, the bounding polygon "hugs" the samples tightly, meaning it has width 3 and height 3.

This would lead to a measured area of 9, while the area derived from the binary image is 16. I just discussed differentiation between discrete/rasterised and continous/geometric measurements with @dietzc and he remotely discussed with @tibuch . I'm fine with these measures being different between binary masks and polygons/meshes. I guess, it's just a matter of being systematic and accurate documentation of possibly appearing differences.

**Comment by ctrueden** [comment]
> while the area derived from the binary image is 16

Well, if we make the 2D polygon operate the same as the 3D mesh, its boundary along x would range from 1.5 to 5.5 (and same for y), yielding a polygon area of 16 as desired (?). It seems like we should do it this way, since it's probably what users will expect, and it's how ImageJ 1.x works.

I also expect that @landinig has an expert opinion here. 😄

**Comment by landinig** [comment]
For me, the best way to deal with this is to think of the pixel being a point, and also for the grid. If you "have" to enlarge it and see a square, then imagine the coordinate system aligned with the centre of "not-a-square" pixels, so in @haesleinhuepf 's example  the width of the grid of a 4x4 square should still 3: count from sample 2 to 3, from 3 to 4 and from 4 to 5.  For **display** purposes you need to think as an underlying grid which addresses the centre of points (so it is half a pixel right and down from the top lef corner of the "not-a-square" pixel. ;-)
We want to keep the integer coordinate system, so the display of ROIs and meshes is what could be changed to be shifted by half a pixel in to the right and down. But then people will complain than when zooming in, a part of a region is outside the ROI...
Sorry @ctrueden a 4x4 pixel pattern has an area of 9, and a pixel has an area of 0 ;-) That is why dealing with 'number of pixels' and region 'area inside the boundary', separately has advantages. Pixel numbers overestimate areas.

**Comment by tpietzsch** [comment]
My personal preference is for the 1.5 ... 5.5 "solution". I think this is safest for example for (re-) rasterizing the polygon/mesh again. For example: Assume we pass the polygon through an affine transform that is basically identity but we have some slight rounding errors (either in the matrix multiplication or in the transform matrix entries). With the -.5 ... +.5 solution we can be rather sure that samples (2,2) ... (5,5) will be rasterized as inside the polygon.

**Comment by haesleinhuepf** [comment]
I like this explanation very much:

> a 4x4 pixel pattern has an area of 9, and a pixel has an area of 0

At the end, I just would like to be able to give biologists a good explanation why measurements are what they are. ;-)

**Comment by imagejan** [comment]
> a pixel has an area of 0 ;-)

If I measure _some_ signal at this pixel sample, and _no_ signal at all neighboring samples, then I'd say the area (or volume) of the object at this position is _anything_ between 0x0=0 (the pixel itself) and 2x2=4 (the neighborhood pixels in case of grid sampling), so why not agree on the most likely area of 1x1=1?

With this in mind, I'd support the -.5 ... +.5 solution. 😄

**Comment by ctrueden** [comment]
See also [this recent discussion on the forum](http://forum.imagej.net/t/polygon-mesh-boundary/3153).
